Sources/Workspace/Workspace.swift

```diff
@@ -2561,168 +2684,208 @@ extension Workspace {
+            fatalError("registry dependencies are supported at this point")
+        }
+    }
 
-    /// Updates the current working checkouts i.e. clone or remove based on the
-    /// provided dependency resolution result.
+    /// Retrieves the given `package` at a given `checkoutState`.
     ///
     /// - Parameters:
-    ///   - updateResults: The updated results from dependency resolution.
-    ///   - diagnostics: The diagnostics engine that reports errors, warnings
-    ///     and notes.
-    ///   - updateBranches: If the branches should be updated in case they're pinned.
-    @discardableResult
-    fileprivate func updateCheckouts(
-        root: PackageGraphRoot,
-        updateResults: [(PackageReference, BoundVersion, ProductFilter)],
-        updateBranches: Bool = false,
-        diagnostics: DiagnosticsEngine
-    ) -> [(PackageReference, PackageStateChange)] {
-        // Get the update package states from resolved results.
-        guard let packageStateChanges = diagnostics.wrap({
-            try computePackageStateChanges(root: root, resolvedDependencies: updateResults, updateBranches: updateBranches)
-        }) else {
-            return []
-        }
-
-        // First remove the checkouts that are no longer required.
-        for (packageRef, state) in packageStateChanges {
-            diagnostics.wrap {
-                switch state {
-                case .added, .updated, .unchanged: break
-                case .removed:
-                    try remove(package: packageRef)
-                }
-            }
-        }
-
-        // Update or clone new packages.
-        for (packageRef, state) in packageStateChanges {
-            diagnostics.wrap {
-                switch state {
-                case .added(let state):
-                    _ = try clone(package: packageRef, requirement: state.requirement, productFilter: state.products)
-                case .updated(let state):
-                    _ = try clone(package: packageRef, requirement: state.requirement, productFilter: state.products)
-                case .removed, .unchanged: break
-                }
-            }
-        }
-
-        // Inform the delegate if nothing was updated.
-        if packageStateChanges.filter({ $0.1 == .unchanged }).count == packageStateChanges.count {
-            delegate?.dependenciesUpToDate()
+    ///   - package: The package to retrieve.
+    ///   - checkoutState: The state to retrieve at.
+    /// - Returns: The path of the local copy of the package.
+    func retrieve(package: PackageReference, at checkoutState: CheckoutState) throws -> AbsolutePath {
+        switch package.kind {
+        case .root, .fileSystem:
+            fatalError("local dependencies are supported")
+        case .localSourceControl, .remoteSourceControl:
+            return try self.checkoutRepository(package: package, at: checkoutState)
+        case .registry:
+            fatalError("registry dependencies are supported at this point")
```

Comment:
@mattt this is where you can plug the registry "fetch" operation. we may want to do something about CheckoutState since its very source control centric. open to ideas